typescript/infra/src/config/agent.ts

```diff
@@ -1,57 +1,188 @@
+import { ValidatorAgentAwsUser } from '../agents/aws';
 
-interface IndexingConfig {
-  from: number;
-  chunk: number;
+// Allows a "default" config to be specified and any per-network overrides.
+interface ChainOverridableConfig<Networks extends ChainName, T> {
+  default: T;
+  chainOverrides?: Partial<ChainSubsetMap<Networks, T>>;
 }
 
-interface AwsConfig {
+// Returns the default config with any overriden values specified for the provided chain.
+export function getChainOverriddenConfig<Networks extends ChainName, T>(
+  overridableConfig: ChainOverridableConfig<Networks, T>,
+  chain: Networks,
+): T {
+  return {
+    ...overridableConfig.default,
+    ...overridableConfig.chainOverrides?.[chain],
+  };
+}
+
+// =====================================
+// =====     Checkpoint Syncer     =====
+// =====================================
+
+// These values are eventually passed to Rust, which expects the values to be camelCase
+export enum CheckpointSyncerType {
+  LocalStorage = 'localStorage',
+  S3 = 's3',
+}
+
+interface LocalCheckpointSyncerConfig {
+  type: CheckpointSyncerType.LocalStorage;
+  path: string;
+}
+
+interface S3CheckpointSyncerConfig {
+  type: CheckpointSyncerType.S3;
+  bucket: string;
   region: string;
 }
 
-interface ProcessorConfig {
-  indexOnly: string[];
-  s3Bucket: string;
+export type CheckpointSyncerConfig =
+  | LocalCheckpointSyncerConfig
+  | S3CheckpointSyncerConfig;
+
+interface MultisigCheckpointSyncerConfig {
+  threshold: number;
+  checkpointSyncers: {
+    [validatorAddress: string]: CheckpointSyncerConfig;
+  };
+}
+
+// =================================
+// =====     Validator Set     =====
+// =================================
+
+// A validator set for a single chain
+interface ValidatorSet {
+  threshold: number;
+  validators: Array<Validator>;
 }
 
-interface RelayerConfig {
-  // How often a relayer should check for new signed checkpoints
-  interval?: number;
+// A validator. This isn't agent-specific configuration, just information
+// on the validator that is enrolled in a validator set.
+interface Validator {
+  address: string;
+  checkpointSyncer: CheckpointSyncerConfig;
 }
 
-interface ValidatorConfig {
-  // How often an validator should check for new updates
-  interval?: number;
-  // How long an validator should wait for relevant state changes afterwords
-  pause?: number;
-  confirmations: number;
+// Validator sets for each network
+export type ChainValidatorSets<Networks extends ChainName> = ChainSubsetMap<
+  Networks,
+  ValidatorSet
+>;
+
+// =================================
+// =====     Relayer Agent     =====
+// =================================
+
+// Incomplete basic relayer agent config
+interface BaseRelayerConfig {
+  // The minimum latency in seconds between two relayed checkpoints on the inbox
+  submissionLatency: number;
+  // The polling interval to check for new checkpoints in seconds
+  pollingInterval?: number;
```

Comment:
Should this be optional?

Reply:
the optional types have default values in the Rust code, but to be explicit with all config values I'll change to make them mandatory